Feature/exam filter

tiago-falves opened the pull request:
Closes #262 

- Added an icon to the exams page
- When the filter icon is clicked a checkbox appears with all exam types
- Only the exams that are checked appear in the exams, including in the favourite cards
- The user choices are stored in AppSharedPreferences, so the data is only recorded when the user chooses to remember the session.
- Modified exam test to include filteredExams in the mock store
- Corrected Covid Exams type

https://user-images.githubusercontent.com/48015828/111355328-8aa97380-8687-11eb-9d0c-bad3fc151139.mp4



# Review checklist
- [ ] Terms and conditions reflect the current change
- [ ] Contains enough appropriate tests
- [ ] Increments version in `pubspec.yaml` (changes `1.0.0+1` to `1.0.0+2` for example)
- [ ] If PR includes UI updates/additions, its description has screenshots
- [ ] Behavior is as expected
- [ ] Clean, well structured code

limwa commented on app_feup/lib/model/entities/exam.dart:
```diff
@@ -93,4 +102,14 @@ class Exam {
+  }
+
+  static getExamTypeLong(String abr) {
+    if (abr == null) return '';
+    final Map<String, String> reversed = _types.map((k, v) => MapEntry(v, k));
+    return reversed[abr];
```
This method can return `null`. Since line 111 avoids returning a null value, I think this is unwanted behaviour. A good fix would be replacing lines 112 and 113 with `return _types.keys.firstWhere((type) => _types[type] == abr, orElse: () => '');`.
Another good fix would be replacing line 113 with `return reversed[abr] ?? '';`.



Personally, I think it would be best to simply return `null` in both situations: when `abr` is `null`  and when `abr` is not a key of `_types`. This would allow the function to be used in other places more easily.

Therefore, the code would look something along the lines of
```suggestion
try {
  return _types.keys.firstWhere((type) => _types[type] == abr));
} catch (e) {
  return null;
}
```

tiago-falves replied:
Done! Thanks!

reeckset commented:
Design feedback:
- Add scrollbar
- Apply button

tiago-falves commented:
- Added styling 
- User does not need to scroll 
- Fixed feature when the session is not persistent 

https://user-images.githubusercontent.com/48015828/111890964-55ce5100-89e6-11eb-93ff-0d0904180a48.mp4

andrefmrocha commented:
Hey! I think this features deserves tests, specially since this is part of an already tested part of the app. Wdyt?

andrefmrocha commented on app_feup/test/integration/src/exams_page_test.dart:
```diff
@@ -78,5 +83,60 @@ void main() {
+      await completerFilter.future;
+
+      await tester.pumpAndSettle();
+      expect(find.byKey(Key(sdisExam.toString())), findsNothing);
+      expect(find.byKey(Key(sopeExam.toString())), findsNothing);
```
I think this integration test should be even more complicated. After what it is doing now, it should open the modal, change the types of exams it wants to see, and then make sure a new exam of that type appears. Wdyt?

andrefmrocha commented on app_feup/lib/controller/local_storage/app_shared_preferences.dart:
```diff
@@ -71,6 +75,30 @@ class AppSharedPreferences {
+    final List<String> storedFilteredExamTypes =
+        prefs.getStringList(filteredExamsTypes);
+
+    final Map<String, bool> examsfiltered = {};
+    if (storedFilteredExamTypes == null) {
+      defaultFilteredExamTypes.forEach((type) => examsfiltered[type] = true);
+      return examsfiltered;
+    }
```
Can we substitue this with a `Map.fromIterable` call?

andrefmrocha commented on app_feup/lib/controller/parsers/parser_exams.dart:
```diff
@@ -3,20 +3,11 @@ import 'package:http/http.dart' as http;
   String getExamSeasonAbbr(String seasonStr) {
-    for (String type in types.keys) {
-      if (seasonStr.contains(type)) return types[type];
+    for (String type in Exam.getExamTypes().keys) {
```
Can we extract the call to `Exam.getExamTypes()` to a variable?

andrefmrocha commented on app_feup/lib/redux/reducers.dart:
```diff
@@ -197,3 +199,7 @@ AppState setLastUserInfoUpdateTime(
 }
+
+AppState setExamFilter(AppState state, SetExamFilter action) {
+  return state.cloneAndUpdateValue('filteredExams', action.filteredExams);
```
Can we put a log message here like the other ones?

andrefmrocha commented on app_feup/test/integration/src/exams_page_test.dart:
```diff
@@ -82,9 +79,83 @@ void main() {
+
+      expect(find.byKey(Key(sdisExam.toString())), findsNothing);
+      expect(find.byKey(Key(sopeExam.toString())), findsNothing);
     });
```
Can we create another test case for this whenever the filters want to display a type of exam that does not exist? I'm scared this could cause unexpected behavior whenever an exam type disappears and would love to have this  covered by the tests. Wdyt?

tiago-falves replied:
This test does not make a lot of sense with this implementation since all the checkboxes are initially checked.
Despite of this, I have created an ExamDoesNotExist test,  in which, even if the Exam-filter map has an unknown type, only the checkboxes that are already know appear.

andrefmrocha replied:
Alright, cool!

andrefmrocha reviewed (changes requested):
Great work! Just some minor tweaks and another test case and it's gtg for me!

andrefmrocha commented:
Don't forget to rebase with develop as well!

andrefmrocha reviewed (dismissed):
:rocket:

reeckset reviewed (commented):
This looks great, including testing. UX-wise, was expecting not to have to click the checkbox itself to toggle it, but rather anywhere on the row. Also, the bug of refreshing seems to be affecting the checkboxes. If you change them and wait a minute they reset back to the original configuration.

reeckset commented:
Will create issue for checkbox UX and approve
